File: python/dijkstra.py
from typing import Dict, List, Tuple
import heapq

# dijkstra algorithm does not work a graph with negative loops

class Dijkstra:
    def __init__(self, edges: List[Tuple[int, int, int]], num_nodes: int) -> None:
        self.edges = edges
        self.num_nodes = num_nodes
        self.graph = self.build_graph()

    def build_graph(self) -> Dict[int, List[Tuple[int, int]]]:
        graph: Dict[int, List[Tuple[int, int]]] = dict()
        for i in range(self.num_nodes):
            graph[i] = []
        for start, end, weight in self.edges:
            graph[start].append((end, weight))
        return graph

    def shortest_path(self, start: int, end: int) -> Tuple[str, int]:
        visited = [False] * len(self.graph)
        prev = [-1] * len(self.graph)
        distances = [10**9] * len(self.graph)
        heap = []
        heapq.heapify(heap)

        # initialize
        distances[start] = 0
        heapq.heappush(heap, (0, start))

        while heap:
            cur_d, cur_idx = heapq.heappop(heap)
            # retrieve the nearest node
            visited[cur_idx] = True
            if cur_idx == end:
                break

            for adj, value in self.graph[cur_idx]:
                next_d = cur_d + value
                # skip visited nodes
                if visited[adj]:
                    continue
                # add new nearest distance and point to heap
                if next_d < distances[adj]:
                    # 注意：在加进的时候更新distance和parent
                    distances[adj] = next_d
                    prev[adj] = cur_idx
                    heapq.heappush(heap, (next_d, adj))
                    
        answer = str(end)
        cur = end
        while cur!=-1:
            cur = prev[cur]
            if cur != -1:
                answer = "{}->{}".format(cur, answer)
        
        return answer, distances[end]


if __name__ == "__main__":
    num_nodes = 9
    edges = [
        (0, 1, 4),
        (0, 7, 8),
        (1, 0, 4),
        (1, 2, 8),
        (1, 7, 11),
        (2, 1, 8),
        (2, 3, 7),
        (2, 5, 4),
        (2, 8, 2),
        (3, 2, 7),
        (3, 4, 9),
        (3, 5, 14),
        (4, 3, 9),
        (4, 5, 10),
        (5, 2, 4),
        (5, 3, 14),
        (5, 4, 10),
        (5, 6, 2),
        (6, 5, 2),
        (6, 7, 1),
        (6, 8, 6),
        (7, 0, 8),
        (7, 1, 11),
        (7, 6, 1),
        (7, 8, 7),
        (8, 2, 2),
        (8, 6, 6),
        (8, 7, 7),
    ]
    instance = Dijkstra(edges, num_nodes)
    start = 1
    end = 1
    path, distance = instance.shortest_path(start, end)
    print("path:{}, distance:{}".format(path, distance))

File: java/dijkstra.java

import java.util.Arrays;
import java.util.PriorityQueue;

public class dijkstra {
    private static int[][][] graph;

    public static void main(String[] args) {
        graph = new int[][][]{{{1,4},{7,8}},{{0,4},{2,8},{7,11}},{{1,8},{3,7},{5,4},{8,2}},{{2,7},{4,9},{5,14}},{{3,9},{5,10}},{{2,4},{3,14},{4,10},{6,2}},{{5,2},{7,1},{8,6}},{{0,8},{1,11},{6,1},{8,7}},{{2,2},{6,6},{7,7}}};
        boolean validateRes = validateGraph(graph);
        System.out.println("the validate result is "+validateRes);
        String path = shortestPath(graph, 1, 8);
        System.out.println(path);
    }

    private static String shortestPath(int[][][] graph, int start, int end) {
        int MAX_VALUE = (int)1e7;
        int n = graph.length;
        int[] distance = new int[n];
        boolean[] visited = new boolean[n];
        int[] prev = new int[n];
        Arrays.fill(distance, MAX_VALUE);
        Arrays.fill(prev, -1);

        // initialization
        // int cur = start;
        distance[start] = 0;
        PriorityQueue<int[]> pq = new PriorityQueue<>((o1,o2)->Double.compare(o1[1], o2[1]));
        pq.offer(new int[]{start, 0});

        while(!pq.isEmpty()) {
            int[] point = pq.poll();
            int cur = point[0];
            visited[cur] = true;
            int[][] adj = graph[cur];
            // update shortest path
            for(int i=0;i<adj.length;i++) {
                int v = adj[i][0];
                int weight = adj[i][1];
                if(!visited[v] && distance[cur]+weight<distance[v]) {
                    distance[v] = distance[cur]+weight;
                    prev[v] = cur;
                    pq.offer(new int[]{v, distance[v]});
                }
            }
        }
        // find the path
        int cur = end;
        String ans = Integer.toString(cur);
        while(cur!=-1) {
            int last = prev[cur];
            if(last!=-1) {
                ans = last+"->"+ans;
            } else {
                ans = "Distance:"+distance[end]+" path:"+ans;
            }
            cur = last;
        }
        return ans;
    }

    private static boolean validateGraph(int[][][] graph) {
        boolean sign = true;
        for(int i=0;i<graph.length;i++) {
            for(int j=0;j<graph[i].length;j++) {
                int verticle = graph[i][j][0];
                int edge = graph[i][j][1];
                boolean checked = false;
                for(int k=0;k<graph[verticle].length;k++) {
                    if(graph[verticle][k][0]==i) {
                        if(graph[verticle][k][1]==edge) {
                            checked = true;
                        } else {
                            checked = true;
                            sign = false;
                        }
                    }
                }
                if(!checked) {
                    sign = false;
                }
            }
        }

        return sign;
    }
}